### PR TITLE
Adds non-zero exit code when the number of arguments are not appropiate

### PR DESCRIPTION
**Pull request by dominoFire**
## Bug
Fixes: https://github.com/NuGet/Home/issues/7178
Regression: No

## Fix
Details: Throw an ExitCodeException when the number of arguments are not correct.

## Testing/Validation
Tests Added: Yes
Validation done: CI Pipepine. 

**Tests Added**

InstallCommand_Failure_WrongArguments
LocalsCommand_Failure_InvalidArguments_HelpMessage

**Tests Modified**

AddCommand_Failure_InvalidArguments_HelpMessage
InitCommand_Failure_InvalidArguments_HelpMessage
LocalsCommand_Success_HelpMessage
LocalsCommand_Failure_InvalidArguments_HelpMessage

### CODE REVIEW

File: src/NuGet.Clients/NuGet.CommandLine/Program.cs [hunk 1]
```diff
@@ -130,6 +130,8 @@ public static int MainCore(string workingDirectory, string[] args)
                     console.WriteLine(LocalizedResourceManager.GetString("InvalidArguments"), commandName);
 
                     p.HelpCommand.ViewHelpForCommand(commandName);
+
+                    throw new ExitCodeException(1);
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/Program.cs]: I would do `return 1;` similar to what's being done in https://github.com/NuGet/NuGet.Client/pull/2505/files#diff-c067a9e9f3d8e725248a12832512a4aaR160

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs [hunk 2]
```diff
@@ -1615,6 +1615,29 @@ public void InstallCommand_AuthenticatedV3WithCredentialSavedInConfig()
...
+            // prepare
+            string[] args = {
+                "install",
+                "mypackage",
+                "-version",
+                "-outputdirectory",
+            };
+
+            // act & assert
+            using(var test_dir = TestDirectory.CreateInTemp())
```
Comment (PatoBeltran) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs]: nit: add a space after `using`. 

Also `testDirectory`, we use camelCase throughout the product. 

Instead of `TestDirectory.CreateInTemp()`, please use `TestDirectory.Create()`. This will create it inside `.test/work` instead of in `%temp%`.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs [hunk 3]
```diff
@@ -1615,6 +1615,29 @@ public void InstallCommand_AuthenticatedV3WithCredentialSavedInConfig()
...
+            {
+                var result = Program.MainCore(test_dir, args);
```
Comment (PatoBeltran) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs]: Instead of using `Program.MainCore` I think is a better idea to use the `CommandRunner`. [Here](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetVerifyCommandTest.cs#L17) is an example of a test using it.

**Comment by PatoBeltran** [comment]
Look at the Add, Init and Locals functional tests that are failing... If you cannot run them just grab me and I can help you out.

File: src/NuGet.Clients/NuGet.CommandLine/Program.cs (hunk 1 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.CommandLine/Program.cs]: Let's try with `return 1;`

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs (hunk 2 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs]: Ok, I applied these on the upcoming commit.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs (hunk 3 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs]: Changed to `CommandRunner`.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs [hunk 4]
```diff
@@ -1199,7 +1199,7 @@ public void NetworkCallCount_RestoreSolutionMultipleSourcesTwoV3()
                     "-Verbosity",
                     "detailed",
                     "-DisableParallelProcessing",
-                    "MSBuildVersion",
+                    "-MSBuildVersion",
```
Comment (PatoBeltran) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs]: Why this changed?

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetAddCommandTests.cs [hunk 5]
```diff
@@ -423,8 +423,8 @@ public void AddCommand_Success_InvalidArguments_HelpMessage(string args)
                 waitForExit: true);
 
             // Assert
-            Util.VerifyResultSuccess(result,
-                "usage: NuGet add <packagePath> -Source <folderBasedPackageSource> [options]");
+            Util.VerifyResultFailure(result,
+                "usage: NuGet add <packagePath> -Source <folderBasedPackageSource> [options]", false);
```
Comment (PatoBeltran) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetAddCommandTests.cs]: please use named parameter for `false`

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInitCommandTests.cs [hunk 6]
```diff
@@ -158,7 +158,10 @@ public void InitCommand_Success_InvalidArguments_HelpMessage(string args)
                 waitForExit: true);
 
             // Assert
-            Util.VerifyResultSuccess(result, "usage: NuGet init <srcPackageSourcePath> <destPackageSourcePath> [options]");
+            // Why assert true ?
```
Comment (PatoBeltran) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInitCommandTests.cs]: Remove comment

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInitCommandTests.cs [hunk 7]
```diff
@@ -158,7 +158,10 @@ public void InitCommand_Success_InvalidArguments_HelpMessage(string args)
...
+            //Util.VerifyResultSuccess(result, "usage: NuGet init <srcPackageSourcePath> <destPackageSourcePath> [options]");
```
Comment (PatoBeltran) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInitCommandTests.cs]: Also remove commented code

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInitCommandTests.cs (hunk 6 above)
Comment (PatoBeltran) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInitCommandTests.cs]: Make sure to change the name of the test since it is now expecting failure instead of success.

`InitCommand_Failure_InvalidArguments_HelpMessage`

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs [hunk 8]
```diff
@@ -1615,6 +1615,35 @@ public void InstallCommand_AuthenticatedV3WithCredentialSavedInConfig()
             }
         }
 
+        /// <summary>
+        /// Unit test that proves non-zero exit code functionality when the number of
+        /// arguments for install command are not appropiate
+        /// </summary>
+        [Fact]
+        public void InstallCommand_WrongArguments()
```
Comment (PatoBeltran) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs]: Consider using a similar name to similar tests: `InstallCommand_Failure_InvalidArguments_HelpMessage`

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs [hunk 9]
```diff
@@ -1615,6 +1615,35 @@ public void InstallCommand_AuthenticatedV3WithCredentialSavedInConfig()
...
+
+            // act & assert
+            using (var testDir = TestDirectory.Create())
+            {
+                var result = CommandRunner.Run(
+                   nugetexe,
+                   testDir,
+                   string.Join(" ", args),
+                   true);
+                Assert.NotEqual(0, result.Item1);
```
Comment (PatoBeltran) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs]: Can you use `Util.VerifyResultSuccess(…)`?

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetLocalsCommandTest.cs [hunk 10]
```diff
@@ -33,6 +32,21 @@ public void LocalsCommand_Success_InvalidArguments_HelpMessage(string args)
             Util.VerifyResultSuccess(result, LocalsHelpStringFragment);
         }
 
+        [Theory]        
+        [InlineData("locals all -list extraArg")]
```
Comment (PatoBeltran) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetLocalsCommandTest.cs:36]: A `Theory` is supposed to be used to run multiple use cases for the same test. Either add more test cases or change this test to be a `Fact`

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs [hunk 11]
```diff
@@ -807,15 +807,20 @@ public static void VerifyResultSuccess(CommandRunnerResult result, string expect
         }
 
         public static void VerifyResultFailure(CommandRunnerResult result,
-                                               string expectedErrorMessage)
+                                               string expectedErrorMessage, bool checkErrorMsgOnStdErr = true)
```
Comment (PatoBeltran) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs]: nit: Add a new line for new parameter

**Review by PatoBeltran — COMMENTED** [review]
Overall is looking good. Great job! Just a couple of small comments.

File: test/TestUtilities/Test.Utility/CommandRunnerResult.cs [hunk 12]
```diff
@@ -9,8 +9,20 @@ namespace NuGet.Test.Utility
     public class CommandRunnerResult
     {
         public Process Process { get; }
+
+        /// <summary>
+        /// Exit Status code
+        /// </summary>
         public int Item1 { get; }
```
Comment (nkolev92) [inline test/TestUtilities/Test.Utility/CommandRunnerResult.cs:19]: Honestly we should use the Item1,2,3 but rather that ExitCode, Output and Errors. 

That could be a separate thing though.

**Review by nkolev92 — APPROVED** [review]

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs (hunk 4 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs]: It seems that, in the past, the test only worked in that specific version of MSBuild. By removing that argument, the test worked well.

File: test/TestUtilities/Test.Utility/CommandRunnerResult.cs (hunk 12 above)
Comment (dominoFire) [reply test/TestUtilities/Test.Utility/CommandRunnerResult.cs:19]: I updated the documentation by adding remarks.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs [hunk 13]
```diff
@@ -1199,8 +1199,6 @@ public void NetworkCallCount_RestoreSolutionMultipleSourcesTwoV3()
                     "-Verbosity",
                     "detailed",
                     "-DisableParallelProcessing",
-                    "MSBuildVersion",
```
Comment (jainaashish) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs:1202]: why is this changed?

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetAddCommandTests.cs [hunk 14]
```diff
@@ -423,8 +423,9 @@ public void AddCommand_Success_InvalidArguments_HelpMessage(string args)
                 waitForExit: true);
 
             // Assert
-            Util.VerifyResultSuccess(result,
-                "usage: NuGet add <packagePath> -Source <folderBasedPackageSource> [options]");
+            Util.VerifyResultFailure(result,
+                "usage: NuGet add <packagePath> -Source <folderBasedPackageSource> [options]",
+                checkErrorMsgOnStdErr: false);
```
Comment (jainaashish) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetAddCommandTests.cs]: as discussed offline, all these error messages should be on std error since you're returning non-zero code. Please confirm that.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs [hunk 15]
```diff
@@ -806,16 +806,31 @@ public static void VerifyResultSuccess(CommandRunnerResult result, string expect
...
+        /// Asserts a non-zero status code and a message.
+        /// </summary>
+        /// <param name="result">An instance of <see cref="CommandRunnerResult"/> with command execution results</param>
+        /// <param name="expectedErrorMessage">A portion of the error message to be sent</param>
+        /// <param name="checkErrorMsgOnStdErr">Check error message on stderr, 
+        /// true by default. If false, it checks the message in stdout</param>
         public static void VerifyResultFailure(CommandRunnerResult result,
-                                               string expectedErrorMessage)
+                                               string expectedErrorMessage,
+                                               bool checkErrorMsgOnStdErr = true)
```
Comment (jainaashish) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/Util.cs]: no need to change here. All failure messages should be asserted on std error

**Review by jainaashish — CHANGES_REQUESTED** [review]
Please check all error messages come through Std Error output and revert the changes for `Util.VerifyResultFailure`

**Review by jainaashish — APPROVED** [review]
👏 great job for your first PR.

**Review by PatoBeltran — APPROVED** [review]

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs (hunk 13 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs:1202]: I noted that in newer versions of msbuild, the MSBuildVersion now is a flag, i.e., `-MSBuildVersion`. If I add it as a flag, the test doesn't work indicating that as an invalid flag. Thus I removed it. I can recover it if you consider appropriate.

Comment (nkolev92) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs:1202]: What do you mean by newer versions of msbuild? 

That switch is on nuget.exe side. 

Were the tests failing?

Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs:1202]: @nkolev92 I didn't fully understand the flag was on the NuGet. The test were failing at the time I was working on this issue. By removing the flag, I was able to run the test with no problems. By installing VS 2015 , you can run the test with the flag but, you need to change it to -MSBuildVersion

**Comment by nkolev92** [comment]
@dominoFire 

Is there any leftover work here? 

Can you address and complete the PR if so?

**Comment by zivkan** [comment]
@dominoFire  Another ping. Is this PR going to be completed/merged?

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs (hunk 4 above)
Comment (nkolev92) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs]: Looking into the test, I don't think we should be using a specific msbuild version. 

NuGet.exe should roll forward seemlessly. 

I'd consider removing it, but you don't have to do it in this PR.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs (hunk 13 above)
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NetworkCallCountTest.cs:1202]: Discussed offline. Modified the test to remove the msbuild flag. We want to remove some platform assumptions.
